pynicotine/users.py

```diff
@@ -116,20 +118,86 @@ def request_set_status(self, status):
     def request_user_stats(self, username):
         core.send_message_to_server(slskmessages.GetUserStats(username))
 
-    def watch_user(self, username):
-        """Tell the server we want to be notified of status/stat updates for a
-        user."""
+    def watch_user(self, username, context=None, is_implicit=False):
+        """Tells the server we want to be notified of status updates for a
+        user.
+
+        context is a string specifying where the user is being watched.
+        The same context must be provided when calling unwatch_user().
+
+        is_implicit is set when receiving status updates from the server
+        without sending a message first. At present, this only happens for
+        users in a joined chat room.
+        """
 
         if self.login_status == slskmessages.UserStatus.OFFLINE:
             return
 
-        if username in self.watched:
+        watched_user = self.watched.get(username)
+
+        if watched_user is None:
+            self.watched[username] = watched_user = WatchedUser(username)
+
+        if not context:
+            context = "unknown"
+
+        if context in watched_user.contexts:
+            return
+
+        if not is_implicit and watched_user.is_implicit:
+            core.send_message_to_server(slskmessages.WatchUser(username))
+            core.send_message_to_server(slskmessages.GetUserStatus(username))  # Get privilege status
+            watched_user.is_implicit = False
+
+        watched_user.contexts.add(context)
+        log.add_conn("Watching user %(user)s in context '%(context)s'. Active contexts: %(contexts)s", {
+            "user": username,
+            "context": context,
+            "contexts": watched_user.contexts
+        })
+
+    def unwatch_user(self, username, context=None):
+        """Tells the server we no longer wish to receive status updates for a
+        user.
+
+        context must be the same as previously provided in watch_user().
+        """
+
+        if username == self.login_username:
+            # We can't unwatch ourselves
+            return
+
+        watched_user = self.watched.get(username)
+
+        if watched_user is None:
+            return
+
+        if not context:
+            context = "unknown"
```

Comment:
If the context is unspecified, then perhaps all contexts ought to be discarded?

Reply:
Too easy to misuse and break things. I'll make the context mandatory, since it's a new function.